pca_ica/extract_cells.m

```diff
@@ -0,0 +1,353 @@
+t = pca_info.movie_frames;
+idx_kept = pca_info.trim.idx_kept;
+
+fprintf('%s: Extracting filters...\n',datestr(now));
+
+U = (pca_filters');
+[N,num_pcs] = size(U);
+norms_U = sqrt(sum(U.^2,2));
+U_norm = bsxfun(@times,U,1./norms_U);
+
+% Brute-force approach for 1-norm computation, use GPU if available
+use_gpu = 0;
+gpu_exists = gpuDeviceCount>0;
+if gpu_exists
+    avail_mem = compute_available_memory(1);    
+    f = max(avail_mem/4 - N*(num_pcs+2),0); % maximum available element size 
+    f = f*mem_occup_scale_GPU;
+    chunk_size = floor(f/ (2*N+num_pcs)); %# of pixels at once
+    use_gpu = chunk_size>=1;
+end    
+    
+if use_gpu %GPU    
+    fprintf('\t \t \t Using GPU to accelarate processing\n');
+    U = gpuArray(U);
+    one_norms = gpuArray(zeros(1,N,'single'));
+    norms_U = gpuArray(norms_U);
+    
+    for i = 1:chunk_size:N
+        fin = min(chunk_size-1,N-i);
+        Q = bsxfun(@times,U(i:i+fin,:),1./norms_U(i:i+fin))';
+        one_norms(i:i+fin) = sum(abs(U*Q),1);
+        clear Q;
+    end
+
+    % Transfer variables back to CPU
+    U = gather(U);
+    one_norms = gather(one_norms);
+    norms_U = gather(norms_U);
+    
+else % CPU
+    avail_mem = compute_available_memory(0);
+    f = avail_mem/4 ; % maximum available element size
+    f = f*mem_occup_scale_CPU;
+    chunk_size = floor(f/ (2*N+num_pcs)); %# of pixels at once
+    
+    one_norms = zeros(1,N);
+    for i = 1:chunk_size:N
+        fin = min(chunk_size-1,N-i);
+        Q = U_norm(i:i+fin,:)';
+        one_norms(i:i+fin) = sum(abs(U*Q),1);
+        clear Q;
+    end
+
+end
+
+% Extract filters recursively using the 1-norm criterion
+idx_possible = 1:N;
+F = zeros(N,max_num);
+inv_quality = zeros(1,max_num);
+acc=0;
+
+while true
+    acc = acc+1;
+    [val,idx_this] = min(one_norms(idx_possible));
+    idx_this = idx_possible(idx_this);
+    sig_this = U*U_norm(idx_this,:)';    
+    correl = sig_this.*(1./norms_U);
+    idx_possible = intersect(find(abs(correl)<corr_thresh),idx_possible);
+    F(:,acc) = sig_this;
+    inv_quality(:,acc) = val;
+    % Termination condition
+    if isempty(idx_possible) || acc==max_num
+        break;
+    end
+    
+end
+
+% Truncate F in case there are less components than max_num
+F = F(:,1:acc);
+inv_quality = inv_quality(1:acc)/sqrt(N);
+
+if is_trimmed % Untrim the pixels
+    F_temp = zeros(h*w,acc);
+    F_temp(idx_kept,:) = F;
+    F = F_temp;
+end
+
+fprintf('\t \t \t Extracted %d potential cells \n',acc);
+fprintf('%s: Cleaning filters and removing duplicates...\n',datestr(now));
+[F,idx] = modify_filters(F,filter_thresh);
+inv_quality = inv_quality(idx);
+
+figure,
+plot(inv_quality,'LineWidth',1.5);
+title('Inverse quality metric for the extracted potential cells','Fontsize',16);
+xlabel('Cell index','Fontsize',16)
+ylabel('Inverse quality(normalized to 1)','Fontsize',16)
+
+fprintf('%s: Loading movie for trace extraction...\n',datestr(now));
+M = load_movie(movie_source);
+
+% Extract time traces
```

Comment:
Is this just least squares given a movie and a set of spatial filters (e.g. basis set)?